src/components/tl/ucp/bcast/bcast_knomial.c

```diff
@@ -0,0 +1,89 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "config.h"
+#include "tl_ucp.h"
+#include "bcast.h"
+#include "core/ucc_progress_queue.h"
+#include "tl_ucp_sendrecv.h"
+#include "utils/ucc_math.h"
+
+#define CALC_DIST(_size, _radix, _dist)                                        \
+    do {                                                                       \
+        _dist = 1;                                                             \
+        while (_dist * _radix < _size) {                                       \
+            _dist *= _radix;                                                   \
+        }                                                                      \
+    } while (0)
+
+ucc_status_t ucc_tl_ucp_bcast_knomial_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task      = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team      = task->team;
+    uint32_t           myrank    = team->rank;
+    uint32_t           team_size = team->size;
+    uint32_t           root      = (uint32_t)task->args.root;
+    int                radix     = task->bcast_kn.radix;
+    uint32_t           vrank     = (myrank - root + team_size) % team_size;
+    uint32_t           dist      = task->bcast_kn.dist;
```

Comment:
myrank, team_size, root, vrank and maybe radix-> ucc_rank_t

Reply:
👍